pkg/resolver/messages.go

```diff
@@ -65,27 +72,75 @@ func (s *Message) buildSearchAddonDisabledQuery(ctx context.Context) {
 	klog.V(3).Infof("Messages Query for managed clusters with Search addon disabled: %s\n", sql)
 }
 
-func (s *Message) messageResults(ctx context.Context) ([]*model.Message, error) {
-	klog.V(2).Info("Resolving Messages()")
+// If user has access to even one search addon disabled cluster, show the search disabled message
+func (s *Message) checkUserAccessToDisabledClusters(ctx context.Context, disabledClusters *map[string]struct{}) bool {
 
-	rows := s.pool.QueryRow(ctx, s.query)
+	for _, cluster := range getKeysStructMap(*disabledClusters) { //rbac.CacheInst.GetDisabledClusters()) {
+		for _, mc := range s.userData.ManagedClusters {
+			klog.Info("checking user access to cluster: ", cluster)
 
+			if mc == cluster {
+				return true
+			}
+		}
+	}
+	return false
+}
+func (s *Message) findSrchAddonDisabledClusters(ctx context.Context) (*map[string]struct{}, error) {
+	disabledClusters := make(map[string]struct{})
+
+	rows, err := s.pool.Query(ctx, s.query)
+	if err != nil {
+		klog.Error("Error fetching SearchAddon disabled cluster results from db ", err)
+		rbac.CacheInst.SetDisabledClusters(disabledClusters, err)
+		return &disabledClusters, err
+	}
+	defer rows.Close()
 	if rows != nil {
-		var count int
-		err := rows.Scan(&count)
-		if err != nil {
-			klog.Errorf("Error %s resolving cluster count for query:%s", err.Error(), s.query)
+		for rows.Next() {
+			var srchAddonDisabledCluster string
+			err := rows.Scan(&srchAddonDisabledCluster)
+			if err != nil {
+				klog.Errorf("Error %s resolving cluster count for query: %s", err.Error(), s.query)
+			}
+			disabledClusters[srchAddonDisabledCluster] = struct{}{}
+
+			//Since cache is not valid, update shared cache with disabled clusters result
+			rbac.CacheInst.SetDisabledClusters(disabledClusters, nil)
 		}
-		if count > 0 {
-			messages := make([]*model.Message, 0)
-			kind := "information"
-			desc := "Search is disabled on some of your managed clusters."
-			message := model.Message{ID: "S20",
-				Kind:        &kind,
-				Description: &desc}
-			messages = append(messages, &message)
-			return messages, nil
+	}
+	return &disabledClusters, err
+}
+func (s *Message) messageResults(ctx context.Context) ([]*model.Message, error) {
+	klog.V(2).Info("Resolving Messages()")
+	var returnDisabledMessage bool
+
+	if rbac.CacheInst.SharedCacheDisabledClustersValid() {
+		klog.V(5).Info("Search Addon DisabledClusters Cache valid")
+		returnDisabledMessage = s.checkUserAccessToDisabledClusters(ctx, rbac.CacheInst.GetDisabledClusters())
+	} else {
+		klog.V(5).Info("DisabledClusters Cache not valid - running query to get search addon disabled clusters")
+
+		if disabledClusters, err := s.findSrchAddonDisabledClusters(ctx); err != nil {
+			klog.Error("Error retrieving Search Addon disabled clusters: ", err)
+			return []*model.Message{}, err
+		} else {
+			returnDisabledMessage = s.checkUserAccessToDisabledClusters(ctx, disabledClusters)
 		}
 	}
```

Comment:
This logic should be part of the cache. 
From here we should call something like `rbac.cacheInst.getDisabledClusters()`
Then inside that function we'll check if we have the data in the cache or if we need to request it.